### PR TITLE
Lzma and zstd

### PR DESCRIPTION
**Pull request by Tinsane**
Refactored a bunch of code related to compression.
Added lzma and zstd compression methods.

### CODE REVIEW

**Comment by x4m** [comment]
Cool! Thanks!
@CyberDem0n Alexander, Hi! As far as I remember you was interested in #97 , can you look into this? Does it bring some more performance on your clusters? Vladimir observes 3x better compression with roughly same CPU consumption. @Tinsane is these numbers correct?

Also, I'll rebase branch on a current master.

File: compression.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	FileExtension() string
+}
+
+type Decompressor interface {
+	Decompress(dst io.Writer, src io.Reader) error
+	FileExtension() string
+}
+
+var Compressors map[string]Compressor
+var Decompressors = []Decompressor{
```
Comment (x4m) [inline compression.go]: Not a show stopper, but different styles of initializations look ugly.

File: compression_test.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+	err = compareDecompressedFileWithInitial(initialFilePath)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+}
+
+func TestSmallFileCompression(t *testing.T) {
+	for _, compressingAlgorithm := range compressingAlgorithms {
```
Comment (x4m) [inline compression_test.go]: How about adding lzop here too? If it is no a lot of work, of course. I know that it is somewhat broken #35 butI still hope to find an example of broken file.... in this case we should leave broken file even in an event of failure.

File: crypter.go [hunk 3]
```diff
@@ -2,7 +2,7 @@ package walg
 
 import "io"
 
-// Crypter is responsible for makeing cryptographical pipeline parts when needed
+// Crypter is responsible for making cryptographical pipeline parts when needed
 type Crypter interface {
```
Comment (x4m) [inline crypter.go]: Currently, we have compressor and decompressor. How do you think, is it viable to split crypter into encryptor and decryptor? They share keyring infrastructure and we have an issue about them #92, maybe keepeing them together for now is a good idea.

File: extract.go [hunk 4]
```diff
@@ -45,36 +46,36 @@ func extractOne(ti TarInterpreter, s io.Reader) error {
...
+	for _, decompressor := range Decompressors {
+		if readerMaker.Format() != decompressor.FileExtension() {
+			continue
 		}
-	} else if readerMaker.Format() == Lz4FileExtension {
-		_, err = DecompressLz4(writeCloser, r)
+		err = decompressor.Decompress(writeCloser, readCloser)
 		if err != nil {
-			return errors.Wrap(err, "ExtractAll: lz4 decompress failed. Is archive encrypted?")
+			return errors.Wrap(err, fmt.Sprintf("ExtractAll: %v decompress failed. Is archive encrypted?", decompressor.FileExtension()))
```
Comment (x4m) [inline extract.go]: How about errors.Wrapf? If it's our package

File: s3_tar_ball.go [hunk 5]
```diff
@@ -1,22 +1,20 @@
...
-	"github.com/aws/aws-sdk-go/aws"
-	"bytes"
 )
 
 // S3TarBall represents a tar file that is
 // going to be uploaded to S3.
 type S3TarBall struct {
-	baseDir          string
 	trim             string
 	bkupName         string
```
Comment (x4m) [inline s3_tar_ball.go]: I'm surprised this didn't trigger you :)

File: s3_tar_ball.go [hunk 6]
```diff
@@ -33,14 +31,14 @@ type S3TarBall struct {
...
-// the form `part_....tar.lz4`.
+// the form `part_....tar.[Compressor file extension]`.
 func (tarBall *S3TarBall) SetUp(crypter Crypter, names ...string) {
 	if tarBall.tarWriter == nil {
 		var name string
 		if len(names) > 0 {
 			name = names[0]
 		} else {
-			name = "part_" + fmt.Sprintf("%0.3d", tarBall.partCount) + ".tar." + Lz4FileExtension
+			name = "part_" + fmt.Sprintf("%0.3d", tarBall.partCount) + ".tar." + tarBall.tarUploader.compressor.FileExtension()
```
Comment (x4m) [inline s3_tar_ball.go]: How about tarFileExtension? If it's not overkill

File: utility.go [hunk 7]
```diff
@@ -142,11 +146,39 @@ func getMaxConcurrency(key string, default_value int) int {
...
+			con = defaultValue
 		} else {
 			con = 10
 		}
 	}
 	return max(con, 1)
 }
+
+// CheckType grabs the file extension from PATH.
+func CheckType(path string) string {
```
Comment (x4m) [inline utility.go]: It's not CheckType() it's GetExtensions()
chances are regex can be part of the global state (static)

**Comment by x4m** [comment]
I'll do one more pass through code tomorrow

File: compression.go (hunk 1 above)
Comment (Tinsane) [reply compression.go]: Fixed

File: s3_tar_ball.go (hunk 5 above)
Comment (Tinsane) [reply s3_tar_ball.go]: Fixed :)

File: extract.go (hunk 4 above)
Comment (Tinsane) [reply extract.go]: Fixed

File: utility.go (hunk 7 above)
Comment (Tinsane) [reply utility.go]: Fixed

File: s3_tar_ball.go (hunk 6 above)
Comment (Tinsane) [reply s3_tar_ball.go]: Fixed

**Comment by x4m** [comment]
I think it's fine to merge now. But we need README update. It's good to have some advice which one user should choose. Please, add your benchmarks and suggestions when user should switch to zstd (network limitations, some spare CPU) and when to lzma (storage and network limitations, a lot of unused CPU).

**Comment by x4m** [comment]
Thanks, pushed.

**Pull request merged**